enhancements/installer/assisted-installer-in-cluster.md

```diff
@@ -0,0 +1,641 @@
+
+1. User creates or modifies a BareMetalAsset on the hub cluster so that its cluster field references the desired ClusterDeployment.
+1. Hive creates a BareMetalHost resource corresponding to the BareMetalAsset and uses it to boot the live ISO.
+1. baremetal-operator uses redfish virtualmedia to boot the live ISO.
+1. Agent running on the host contacts assisted service, resulting in an AssistedInstallAgent resource being created.
+1. Assisted Installer walks the agent through discovery and validation workflows.
+1. When the agent is in a ready state, hive or ACM initiates provisioning, referencing the target cluster's ignition.
+
+#### Create Cluster
+
+This scenario takes place on a hub cluster where hive and possible ACM are
+present. This scenario does not include centralized machine management.
+
+1. User creates BareMetalAssets containing BMC credentials.
+1. User creates AssistedInstallCluster resource describing their desired cluster.
+1. User creates ClusterDeployment describing a new cluster. The Platform section indicates that metal3 should be used to boot hosts with a live ISO. A new field references the AssistedInstallCluster resource.
+1. User associates some BareMetalAssets with the ClusterDeployment.
+1. Hive creates BareMetalHost resources corresponding to each BareMetalAsset, and uses them to boot the live ISO.
+1. baremetal-operator uses redfish virtualmedia to boot the live ISO.
+1. Agent running on each host contacts assisted service, resulting in an AssistedInstallAgent resource being created.
+1. Assisted Installer walks each agent through discovery and validation workflows.
+1. When each agent is in a ready state, hive or ACM initiates cluster creation with the AssistedInstallCluster.
+
+### Risks and Mitigations
+
+What are the risks of this proposal and how do we mitigate. Think broadly. For example, consider both security and how this will impact the larger OKD
+ecosystem.
+
+How will security be reviewed and by whom? How will UX be reviewed and by whom?
+
+Consider including folks that also work outside your immediate sub-project.
+
+## Design Details
+
+### Open Questions
+
+#### Can we use sqlite?
+
+It would be advantageous to use sqlite for the database. The [gorm library
+supports sqlite](https://gorm.io/docs/connecting_to_the_database.html#SQLite),
+but it is not clear if assisted service is compatible. In particular, the [use
+of FOR
+UPDATE](https://github.com/openshift/assisted-service/blob/e70af7dcf59763ee6c697fb409887f00ab5540f5/pkg/transaction/transaction.go#L8)
+might be problematic.
+
+#### baremetal-operator watching multiple namespaces?
+
+When utilizing baremetal-operator on a hub cluster to boot the discovery ISO on
+hosts, should we be creating those BareMetalHost resources in a separate
+namespace from those that are associated with the hub cluster itself?
+
+What work is involved in having BMO watch additional namespaces?
```

Comment:
The BMO will already watch multiple namespaces. That's how upstream users run it today. The CBO configures our copy to only watch a single namespace today using a command line flag.

If the hub cluster has its own copy of metal3, the conflict will be between that instance (looking at the openshift-machine-api namespace) and the one used by hive (looking at all namespaces, including openshift-machine-api).

I'm not aware of any way to make the controller-runtime code ignore a namespace, so if we want the BMO to do that we will have to add special code to the controller.